[Clang][OpenMP][LoopTransformations] Add support for "#pragma omp fuse" loop transformation directive and "looprange" clause

eZWALT opened the pull request:
This pull request introduces full support for the #pragma omp fuse directive, as specified in the OpenMP 6.0 specification, along with initial support for the looprange clause in Clang.

To enable this functionality, infrastructure for the Loop Sequence construct, also new in OpenMP 6.0, has been implemented. Additionally, a minimal code skeleton has been added to Flang to ensure compatibility and avoid integration issues, although a full implementation in Flang is still pending.

https://www.openmp.org/wp-content/uploads/OpenMP-API-Specification-6-0.pdf  

@alexey-bataev @Meinersbur

P.S. As a follow-up to this loop transformation work, I'm currently preparing a patch that implements the "#pragma omp split" directive, also introduced in OpenMP 6.0.

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -14175,27 +14222,350 @@ bool SemaOpenMP::checkTransformableLoopNest(
+  unsigned NestedLoopCount = 0;
+
+public:
+  explicit NestedLoopCounterVisitor() {}
```
```suggestion
  explicit NestedLoopCounterVisitor() = default;
```

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -14175,27 +14222,350 @@ bool SemaOpenMP::checkTransformableLoopNest(
+      return true;
+
+    // Only recurse into CompoundStmt (block {}) and loop bodies
+    if (isa<CompoundStmt>(S) || isa<ForStmt>(S) || isa<CXXForRangeStmt>(S)) {
+      return DynamicRecursiveASTVisitor::TraverseStmt(S);
+    }
```
```suggestion
    if (isa<CompoundStmt, ForStmt, CXXForRangeStmt>(S))
      return DynamicRecursiveASTVisitor::TraverseStmt(S);
```

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -14175,27 +14222,350 @@ bool SemaOpenMP::checkTransformableLoopNest(
+  // Helper Lambda to handle storing initialization and body statements for both
+  // ForStmt and CXXForRangeStmt and checks for any possible mismatch between
+  // induction variables types
+  auto storeLoopStatements = [&OriginalInits, &ForStmts, &BaseInductionVarType,
+                              this, &Context](Stmt *LoopStmt) {
```
```suggestion
  auto StoreLoopStatements = [&](Stmt *LoopStmt) {
```

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -14175,27 +14222,350 @@ bool SemaOpenMP::checkTransformableLoopNest(
+    }
+  };
+
+  // Helper lambda functions to encapsulate the processing of different
```
Use `\\\`

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -14175,27 +14222,350 @@ bool SemaOpenMP::checkTransformableLoopNest(
+  // derivations of the canonical loop sequence grammar
+  //
+  // Modularized code for handling loop generation and transformations
+  auto analyzeLoopGeneration = [&storeLoopStatements, &LoopHelpers,
+                                &OriginalInits, &TransformsPreInits,
+                                &LoopCategories, &LoopSeqSize, &NumLoops, Kind,
+                                &TmpDSA, &ForStmts, &Context,
+                                &LoopSequencePreInits, this](Stmt *Child) {
```
```suggestion
  auto AnalyzeLoopGeneration = [&](Stmt *Child) {
```

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -14175,27 +14222,350 @@ bool SemaOpenMP::checkTransformableLoopNest(
+                                &LoopCategories, &LoopSeqSize, &NumLoops, Kind,
+                                &TmpDSA, &ForStmts, &Context,
+                                &LoopSequencePreInits, this](Stmt *Child) {
+    auto LoopTransform = dyn_cast<OMPLoopTransformationDirective>(Child);
```
```suggestion
    auto *LoopTransform = dyn_cast<OMPLoopTransformationDirective>(Child);
```

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -14175,27 +14222,350 @@ bool SemaOpenMP::checkTransformableLoopNest(
+        LoopSeqSize += NumGeneratedLoopNests;
+        NumLoops += NumGeneratedLoops;
+        return true;
+      }
+      // Unroll full (0 loops produced)
+      else {
```
```suggestion
      } else {
        // Unroll full (0 loops produced)
```

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -14175,27 +14222,350 @@ bool SemaOpenMP::checkTransformableLoopNest(
+      return false;
+    }
+    // Loop transformatons such as split or loopranged fuse
+    else if (NumGeneratedLoopNests > 1) {
```
```suggestion
   if (NumGeneratedLoopNests > 1) {
```

alexey-bataev replied:
Coding standard, there should be `else` after a substament with the return

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -14175,27 +14222,350 @@ bool SemaOpenMP::checkTransformableLoopNest(
+                                 LoopCategories, Context, Kind);
+    }
+    // Vast majority: (Tile, Unroll, Stripe, Reverse, Interchange, Fuse all)
+    else {
```
```suggestion
```

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -14175,27 +14222,350 @@ bool SemaOpenMP::checkTransformableLoopNest(
+  };
+
+  // Modularized code for handling regular canonical loops
+  auto analyzeRegularLoop = [&storeLoopStatements, &LoopHelpers, &OriginalInits,
+                             &LoopSeqSize, &NumLoops, Kind, &TmpDSA,
+                             &LoopCategories, this](Stmt *Child) {
```
```suggestion
  auto AnalyzeRegularLoop = [&](Stmt *Child) {
```

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -14175,27 +14222,350 @@ bool SemaOpenMP::checkTransformableLoopNest(
+  };
+
+  // Helper functions to validate canonical loop sequence grammar is valid
+  auto isLoopSequenceDerivation = [](auto *Child) {
```
```suggestion
  auto IsLoopSequenceDerivation = [](auto *Child) {
```

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -14175,27 +14222,350 @@ bool SemaOpenMP::checkTransformableLoopNest(
+
+  // Helper functions to validate canonical loop sequence grammar is valid
+  auto isLoopSequenceDerivation = [](auto *Child) {
+    return isa<ForStmt>(Child) || isa<CXXForRangeStmt>(Child) ||
+           isa<OMPLoopTransformationDirective>(Child);
```
```suggestion
    return isa<ForStmt, CXXForRangeStmt, OMPLoopTransformationDirective>(Child);
```

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -14175,27 +14222,350 @@ bool SemaOpenMP::checkTransformableLoopNest(
+    return isa<ForStmt>(Child) || isa<CXXForRangeStmt>(Child) ||
+           isa<OMPLoopTransformationDirective>(Child);
+  };
+  auto isLoopGeneratingStmt = [](auto *Child) {
```
```suggestion
  auto IsLoopGeneratingStmt = [](auto *Child) {
```

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -14175,27 +14222,350 @@ bool SemaOpenMP::checkTransformableLoopNest(
+    // Regular loop sequence handling
+    if (isLoopSequenceDerivation(Child)) {
+      if (isLoopGeneratingStmt(Child)) {
+        if (!analyzeLoopGeneration(Child)) {
+          return false;
+        }
```
```suggestion
        if (!analyzeLoopGeneration(Child))
          return false;
```

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -14175,27 +14222,350 @@ bool SemaOpenMP::checkTransformableLoopNest(
+        // analyzeLoopGeneration updates Loop Sequence size accordingly
+
+      } else {
+        if (!analyzeRegularLoop(Child)) {
+          return false;
+        }
```
```suggestion
        if (!analyzeRegularLoop(Child))
          return false;
```

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -14175,27 +14222,350 @@ bool SemaOpenMP::checkTransformableLoopNest(
+  if (!analyzeLoopSequence(AStmt, LoopSeqSize, NumLoops, LoopHelpers, ForStmts,
+                           OriginalInits, TransformsPreInits,
+                           LoopSequencePreInits, LoopCategories, Context,
+                           Kind)) {
+    return false;
+  }
```
```suggestion
                           Kind))
    return false;
```

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -15451,6 +15819,500 @@ StmtResult SemaOpenMP::ActOnOpenMPInterchangeDirective(
+  CaptureVars CopyTransformer(SemaRef);
+
+  // Ensure the structured block is not empty
+  if (!AStmt) {
+    return StmtError();
+  }
```
```suggestion
  if (!AStmt)
    return StmtError();
```

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -15451,6 +15819,500 @@ StmtResult SemaOpenMP::ActOnOpenMPInterchangeDirective(
+  if (!checkTransformableLoopSequence(OMPD_fuse, AStmt, LoopSeqSize, NumLoops,
+                                      LoopHelpers, LoopStmts, OriginalInits,
+                                      TransformsPreInits, LoopSequencePreInits,
+                                      LoopCategories, Context)) {
+    return StmtError();
+  }
```
```suggestion
                                      LoopCategories, Context))
    return StmtError();
```

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -15451,6 +15819,500 @@ StmtResult SemaOpenMP::ActOnOpenMPInterchangeDirective(
+  // expressions. Generates both the variable declaration and the corresponding
+  // initialization statement.
+  auto CreateHelperVarAndStmt =
+      [&SemaRef = this->SemaRef, &Context, &CopyTransformer,
+       &IVType](Expr *ExprToCopy, const std::string &BaseName, unsigned I,
```
```suggestion
      [&, &SemaRef = SemaRef](Expr *ExprToCopy, const std::string &BaseName, unsigned I,
```

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -15451,6 +15819,500 @@ StmtResult SemaOpenMP::ActOnOpenMPInterchangeDirective(
+  // Transformations that apply this concept: Loopranged Fuse, Split
+  if (!LoopSequencePreInits.empty()) {
+    for (const auto &LTPreInits : LoopSequencePreInits) {
+      if (!LTPreInits.empty()) {
+        llvm::append_range(PreInits, LTPreInits);
+      }
```
```suggestion
      if (!LTPreInits.empty())
        llvm::append_range(PreInits, LTPreInits);
```

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -15451,6 +15819,500 @@ StmtResult SemaOpenMP::ActOnOpenMPInterchangeDirective(
+      // Order matters: pre-inits may define variables used in the original
+      // inits such as upper bounds...
+      auto TransformPreInit = TransformsPreInits[TransformIndex++];
+      if (!TransformPreInit.empty()) {
+        llvm::append_range(PreInits, TransformPreInit);
+      }
```
```suggestion
      if (!TransformPreInit.empty())
        llvm::append_range(PreInits, TransformPreInit);
```

eZWALT commented on clang/lib/CodeGen/CGExpr.cpp:
```diff
@@ -3223,6 +3223,8 @@ LValue CodeGenFunction::EmitDeclRefLValue(const DeclRefExpr *E) {
 
     // No other cases for now.
     } else {
+      llvm::dbgs() << "THE DAMN DECLREFEXPR HASN'T BEEN ENTERED IN LOCALDECLMAP\n";
+      VD->dumpColor();
```
Oops, thanks!

eZWALT commented on clang/lib/CodeGen/CodeGenFunction.h:
```diff
@@ -5378,6 +5379,10 @@ class CodeGenFunction : public CodeGenTypeCache {
 
   /// Set the address of a local variable.
   void setAddrOfLocalVar(const VarDecl *VD, Address Addr) {
+    if (LocalDeclMap.count(VD)) {
+      llvm::errs() << "Warning: VarDecl already exists in map: ";
+      VD->dumpColor(); 
+    }
```
Oops, thanks!

eZWALT commented on clang/include/clang/AST/StmtOpenMP.h:
```diff
@@ -5790,7 +5805,11 @@ class OMPReverseDirective final : public OMPLoopTransformationDirective {
-                                       EndLoc, 1) {}
+                                       EndLoc, 1) {
+    // Reverse produces a single top-level canonical loop nest
+    setNumGeneratedLoops(1);
```
Okey do i make another pull request just for the loop corrections though?

alexey-bataev replied:
Yes

eZWALT commented on clang/include/clang/AST/StmtOpenMP.h:
```diff
@@ -962,6 +962,9 @@ class OMPLoopTransformationDirective : public OMPLoopBasedDirective {
   unsigned NumGeneratedLoops = 0;
+  /// Number of top level canonical loop nests generated by this loop
+  /// transformation
+  unsigned NumGeneratedLoopNests = 0;
```
Maybe the name is a bit unfortunate and could be improved, but they are 2 completely different fields conceptually. This top level loops are the ones actually managed by loop Sequence constructs like fuse and the upcoming split. A loop sequence contains loops which may contain several inner nestes loops, but these should not be taken into account for performing fusion or splitting.  This was not taken into account originally due to all transformations having a fixed number of generated top level nests (1). However fuse or split may generate several loop nests with inner nested loops.

eZWALT replied:
Note that unroll is an exception, it could have 0 or 1 but it coincides perfectly with the original number of loops .

alexey-bataev replied:
The question is how it is used. I did not see it is being read anywhere

eZWALT replied:
This distinction is indeed important and actively used in `SemaOpenMP.cpp` file, particularly within the `AnalyzeLoopSequence` function (starting at line 14284). For example, it's referenced in lines 14344 and 14364 to differentiate between specific loop transformations.

eZWALT commented:
I want to notify that the following week I'll be unavailable, so expect this patch to be updated on the 20th of May. Thanks for the feedback @alexey-bataev

eZWALT commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -11516,6 +11516,21 @@ def note_omp_implicit_dsa : Note<
+  "loop after '#pragma omp %0' is not in canonical form">;
+def err_omp_not_a_loop_sequence : Error < 
+  "statement after '#pragma omp %0' must be a loop sequence containing canonical loops or loop-generating constructs">;
+def err_omp_empty_loop_sequence : Error <
+  "loop sequence after '#pragma omp %0' must contain at least 1 canonical loop or loop-generating construct">;
+def err_omp_invalid_looprange : Error <
+  "loop range in '#pragma omp %0' exceeds the number of available loops: "
+  "range end '%1' is greater than the total number of loops '%2'">;
```
The two errors serve different purposes:

1. The first is triggered when the loop sequence inside a fusion construct (full or ranged) contains no loops.
2. The second is specific to loopranged fusion and reports when the specified range exceeds the number of available loops.

While the first case is technically a subset of the second, they occur in different contexts. Keeping both improves clarity and helps users better understand the issue. That said, I’m open to refactoring if you'd prefer a single, more general diagnostic, though it may reduce the precision of the error messages.

eZWALT commented on clang/include/clang/Sema/SemaOpenMP.h:
```diff
@@ -1480,6 +1493,108 @@ class SemaOpenMP : public SemaBase {
+      Stmt *LoopSeqStmt, unsigned &LoopSeqSize, unsigned &NumLoops,
+      SmallVectorImpl<OMPLoopBasedDirective::HelperExprs> &LoopHelpers,
+      SmallVectorImpl<Stmt *> &ForStmts,
+      SmallVectorImpl<SmallVector<Stmt *, 0>> &OriginalInits,
+      SmallVectorImpl<SmallVector<Stmt *, 0>> &TransformsPreInits,
+      SmallVectorImpl<SmallVector<Stmt *, 0>> &LoopSequencePreInits,
```
it is correct though, in order to be consistent with the rest of the codebase, there are several instances of this SmallVector<Stmt*, 0> being propagated (In checkTransformableLoopNest, ActOnOpenMPUnroll, ActOnOpenMPStripe...) therefore i either modify all of these instances or i keep it as it is.

eZWALT replied:
At the end i've decided to modify all of them and avoid this rigid fixed logic

eZWALT commented:
Gentle-ping, I'm not sure if GitHub has notified you of the comments :) @alexey-bataev

alexey-bataev commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -11516,6 +11516,21 @@ def note_omp_implicit_dsa : Note<
 def err_omp_loop_var_dsa : Error<
   "loop iteration variable in the associated loop of 'omp %1' directive may not be %0, predetermined as %2">;
+def warn_omp_different_loop_ind_var_types : Warning <
+  "loop sequence following '#pragma omp %0' contains induction variables of differing types: %1 and %2">,
```
What stops you from promoting to a larger type?

eZWALT replied:
Nothing actually, it was one of the 3 ideas i had to tackle this problem and if you prefer i will follow this approach (Using uint of 64 bits for example, however your comment left me thinking if NumIterations cause any trouble, i think it can't but i'm not 100% sure to be honest.

eZWALT replied:
@alexey-bataev

eZWALT commented:
gentle ping @alexey-bataev

Meinersbur commented on openmp/runtime/test/transform/fuse/intfor.c:
```diff
@@ -0,0 +1,50 @@
+  printf("do\n");
+#pragma omp fuse
+  {
+    for (int i = 5; i <= 25; i += 5)
+      printf("i=%d\n", i);
+    for (int j = 10; j < 100; j += 10)
+      printf("j=%d\n", j);
+    for (int k = 10; k > 0; --k)
+      printf("k=%d\n", k);
```
Nice test.

I hope these end2end tests were as useful for you as they were for me

Meinersbur commented on openmp/runtime/test/transform/fuse/parallel-wsloop-collapse-intfor.c:
```diff
@@ -0,0 +1,45 @@
+#endif /* HEADER */
+
+// CHECK:      do
+// CHECK: i=0 j=0
```
```suggestion
// CHECK-NEXT: i=0 j=0
```

Meinersbur reviewed (commented):
Thanks for the work and sorry I could not have a look into it earlier

Meinersbur reviewed (commented):
Thanks for the work and sorry I could not have a look into it earlier

eZWALT commented:
> Thanks for the work and sorry I could not have a look into it earlier

Nono, thank you for your time and guidance! Between today and tomorrow i'll upload the updated version, thanks for taking the time to improve it and the nitpicks!

eZWALT commented on clang/lib/CodeGen/CGExpr.cpp:
```diff
@@ -3256,9 +3256,8 @@ LValue CodeGenFunction::EmitDeclRefLValue(const DeclRefExpr *E) {
 
     // No other cases for now.
-    } else {
+    } else
       llvm_unreachable("DeclRefExpr for Decl not entered in LocalDeclMap?");
-    }
```
This should not be in the pr thanks!

eZWALT replied:
> This shouldn't be inside the pr thanks!

eZWALT commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -14189,10 +14196,49 @@ StmtResult SemaOpenMP::ActOnOpenMPTargetTeamsDistributeSimdDirective(
 bool SemaOpenMP::checkTransformableLoopNest(
     OpenMPDirectiveKind Kind, Stmt *AStmt, int NumLoops,
     SmallVectorImpl<OMPLoopBasedDirective::HelperExprs> &LoopHelpers,
-    Stmt *&Body, SmallVectorImpl<SmallVector<Stmt *, 0>> &OriginalInits) {
```
The same i told you before in the other comment, i refactored it to match the suggestion of alexey.

eZWALT commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -15499,6 +15836,496 @@ StmtResult SemaOpenMP::ActOnOpenMPInterchangeDirective(
+    auto [IVVD, IVDStmt] =
+        CreateHelperVarAndStmt(LoopHelpers[I].IterationVarRef, "iv", J);
+
+    if (!LBVD || !STVD || !NIVD || !IVVD)
+      assert(LBVD && STVD && NIVD && IVVD &&
+             "OpenMP Fuse Helper variables creation failed");
```
wow

eZWALT commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -15499,6 +15836,496 @@ StmtResult SemaOpenMP::ActOnOpenMPInterchangeDirective(
+  //    }
+
+  // 1. Create the initialized fuse index
+  const std::string IndexName = Twine(".omp.fuse.index").str();
```
Thanks for spotting!

alexey-bataev commented on clang/include/clang/AST/OpenMPClause.h:
```diff
@@ -1143,6 +1143,101 @@ class OMPFullClause final : public OMPNoChildClause<llvm::omp::OMPC_full> {
+  SourceLocation FirstLoc, CountLoc;
+
+  /// Number of looprange arguments (always 2: first, count)
+  unsigned NumArgs = 2;
```
Is it always constant? In this case you don't need to use tail allocation, use fixed array

alexey-bataev commented on clang/include/clang/AST/OpenMPClause.h:
```diff
@@ -1143,6 +1143,101 @@ class OMPFullClause final : public OMPNoChildClause<llvm::omp::OMPC_full> {
+  /// Get looprange 'count' expression
+  Expr *getCount() const { return getArgs()[1]; }
+
+  /// Set looprange 'first' expression
+  void setFirst(Expr *E) { getArgs()[0] = E; }
+
+  /// Set looprange 'count' expression
+  void setCount(Expr *E) { getArgs()[1] = E; }
+
```
These functions should be private

alexey-bataev commented on clang/include/clang/AST/OpenMPClause.h:
```diff
@@ -1143,6 +1143,101 @@ class OMPFullClause final : public OMPNoChildClause<llvm::omp::OMPC_full> {
+  /// Set looprange 'count' expression
+  void setCount(Expr *E) { getArgs()[1] = E; }
+
+  MutableArrayRef<Expr *> getArgs() {
+    return MutableArrayRef<Expr *>(getTrailingObjects<Expr *>(), NumArgs);
+  }
+  ArrayRef<Expr *> getArgs() const {
+    return ArrayRef<Expr *>(getTrailingObjects<Expr *>(), NumArgs);
+  }
```
Doubt you need  these functions

alexey-bataev commented on clang/include/clang/Sema/SemaOpenMP.h:
```diff
@@ -1479,7 +1492,109 @@ class SemaOpenMP : public SemaBase {
-      Stmt *&Body, SmallVectorImpl<SmallVector<Stmt *, 0>> &OriginalInits);
+      Stmt *&Body, SmallVectorImpl<SmallVector<Stmt *>> &OriginalInits);
+
+  /// @brief Categories of loops encountered during semantic OpenMP loop
```
You can remove `@brief` tag

alexey-bataev commented on clang/include/clang/Sema/SemaOpenMP.h:
```diff
@@ -1479,7 +1492,109 @@ class SemaOpenMP : public SemaBase {
+    /// Represents a standard canonical loop nest found in the
+    /// original source code or an intact loop after transformations
+    /// (i.e Post/Pre loops of a loopranged fusion)
+    RegularLoop,
```
Better to commit support for these categories in separate patches, i.e. initially commit support only for single subclass, then add another one and then the third one in separates patches

alexey-bataev commented on clang/include/clang/Sema/SemaOpenMP.h:
```diff
@@ -1479,7 +1492,109 @@ class SemaOpenMP : public SemaBase {
+  ///                       successfully analyzed.
+  /// \param Context
+  /// \return Whether there was an absence of errors or not
+  bool checkTransformableLoopSequence(
+      OpenMPDirectiveKind Kind, Stmt *AStmt, unsigned &LoopSeqSize,
+      unsigned &NumLoops,
+      SmallVectorImpl<OMPLoopBasedDirective::HelperExprs> &LoopHelpers,
+      SmallVectorImpl<Stmt *> &ForStmts,
+      SmallVectorImpl<SmallVector<Stmt *>> &OriginalInits,
+      SmallVectorImpl<SmallVector<Stmt *>> &TransformsPreInits,
+      SmallVectorImpl<SmallVector<Stmt *>> &LoopSequencePreInits,
+      SmallVectorImpl<OMPLoopCategory> &LoopCategories, ASTContext &Context);
```
There are too many arguments, better to pack them into a struct

alexey-bataev replied:
Not addressed

rofirrim replied:
Done. I think the structs make a bit clearer how @eZWALT approached the analysis here.

alexey-bataev commented on clang/lib/AST/OpenMPClause.cpp:
```diff
@@ -1024,6 +1024,29 @@ OMPPartialClause *OMPPartialClause::CreateEmpty(const ASTContext &C) {
+OMPLoopRangeClause::Create(const ASTContext &C, SourceLocation StartLoc,
+                           SourceLocation LParenLoc, SourceLocation FirstLoc,
+                           SourceLocation CountLoc, SourceLocation EndLoc,
+                           ArrayRef<Expr *> Args) {
```
Better to pass 2 separate arguments for each expressions, not array

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -14189,10 +14194,48 @@ StmtResult SemaOpenMP::ActOnOpenMPTargetTeamsDistributeSimdDirective(
+/// Tries to recursively cast `t` to one of the given types and invokes `f` if
+/// successful.
+///
+/// @tparam Class The first type to check.
```
Use `\param` etc. style, this is common style for doxygen

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -14189,10 +14194,48 @@ StmtResult SemaOpenMP::ActOnOpenMPTargetTeamsDistributeSimdDirective(
 }
 
+/// Overloaded base case function
+template <typename T, typename F> static bool tryHandleAs(T *t, F &&) {
```
```suggestion
template <typename T, typename F> static bool tryHandleAs(T *, F &&) {
```

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -14189,10 +14194,48 @@ StmtResult SemaOpenMP::ActOnOpenMPTargetTeamsDistributeSimdDirective(
+/// @return `true` if `t` matched any type and `f` was called, otherwise
+/// `false`.
+template <typename Class, typename... Rest, typename T, typename F>
+static bool tryHandleAs(T *t, F &&f) {
```
```suggestion
static bool tryHandleAs(T *Type, F &&Func) {
```

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -14189,10 +14194,48 @@ StmtResult SemaOpenMP::ActOnOpenMPTargetTeamsDistributeSimdDirective(
+/// `false`.
+template <typename Class, typename... Rest, typename T, typename F>
+static bool tryHandleAs(T *t, F &&f) {
+  if (Class *c = dyn_cast<Class>(t)) {
```
```suggestion
  if (Class *C = dyn_cast<Class>(t)) {
```

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -14219,27 +14262,320 @@ bool SemaOpenMP::checkTransformableLoopNest(
+///     }
+/// }
+/// Result: Loop 'i' contains 2 loops, Loop 'r' also contains 2 loops
+class NestedLoopCounterVisitor : public DynamicRecursiveASTVisitor {
```
```suggestion
class NestedLoopCounterVisitor final: public DynamicRecursiveASTVisitor {
```

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -14219,27 +14262,320 @@ bool SemaOpenMP::checkTransformableLoopNest(
+    }
+    // Handle loop transformations with multiple loop nests
+    // Unroll full
+    if (NumGeneratedLoopNests <= 0) {
```
Can it be less than 0?

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -14219,27 +14262,320 @@ bool SemaOpenMP::checkTransformableLoopNest(
+                                 LoopHelpers, ForStmts, OriginalInits,
+                                 TransformsPreInits, LoopSequencePreInits,
+                                 LoopCategories, Context, Kind);
+    } else {
```
Remove else

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -15499,6 +15836,496 @@ StmtResult SemaOpenMP::ActOnOpenMPInterchangeDirective(
+
+  // Firstly we need to update TransformIndex to match the begining of the
+  // looprange section
+  for (unsigned int I = 0; I < FirstVal - 1; ++I) {
```
```suggestion
  for (unsigned I : llvm::seq<unsigned>(FirstVal - 1)) {
```

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
Better to use ArrayRef

rofirrim commented:
Hi all, @eZWALT is changing jobs and kindly asked me if I could finish on his behalf. I plan to go through all the items of feedback and then rebase against main.

alexey-bataev commented on clang/include/clang/AST/OpenMPClause.h:
```diff
@@ -1143,6 +1143,83 @@ class OMPFullClause final : public OMPNoChildClause<llvm::omp::OMPC_full> {
+  static OMPLoopRangeClause *
+  Create(const ASTContext &C, SourceLocation StartLoc, SourceLocation LParenLoc,
+         SourceLocation FirstLoc, SourceLocation CountLoc,
+         SourceLocation EndLoc, Expr* First, Expr* Count);
```
Formatting?

alexey-bataev reviewed (commented):
Some previous comments are not addressed yet

alexey-bataev commented on clang/include/clang-c/Index.h:
```diff
@@ -2162,6 +2162,10 @@ enum CXCursorKind {
 
+  /** OpenMP fuse directive
+   */
+  CXCursor_OMPFuseDirective = 318,
```
```suggestion
  CXCursor_OMPFuseDirective = 311,
```

alexey-bataev commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -11664,6 +11664,17 @@ def note_omp_implicit_dsa : Note<
   "implicitly determined as %0">;
 def err_omp_loop_var_dsa : Error<
   "loop iteration variable in the associated loop of 'omp %1' directive may not be %0, predetermined as %2">;
+def err_omp_not_canonical_loop : Error<
+  "loop after '#pragma omp %0' is not in canonical form">;
```
Unused

rofirrim replied:
They're in `clang/lib/Sema/SemaOpenMP.cpp`. Github is not rendering the diff by default.

alexey-bataev replied:
Why are previous error messages not enough? Why need a new one?

alexey-bataev commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -11664,6 +11664,17 @@ def note_omp_implicit_dsa : Note<
   "loop iteration variable in the associated loop of 'omp %1' directive may not be %0, predetermined as %2">;
+def err_omp_not_canonical_loop : Error<
+  "loop after '#pragma omp %0' is not in canonical form">;
+def err_omp_not_a_loop_sequence : Error<
```
Unused

rofirrim replied:
`clang/lib/Sema/SemaOpenMP.cpp`

alexey-bataev commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -11664,6 +11664,17 @@ def note_omp_implicit_dsa : Note<
+  "loop after '#pragma omp %0' is not in canonical form">;
+def err_omp_not_a_loop_sequence : Error<
+  "statement after '#pragma omp %0' must be a loop sequence containing canonical loops or loop-generating constructs">;
+def err_omp_empty_loop_sequence : Error<
+  "loop sequence after '#pragma omp %0' must contain at least 1 canonical loop or loop-generating construct">;
+def err_omp_invalid_looprange : Error<
+  "looprange clause selects loops from %1 to %2 but this exceeds the number of loops (%3) in the loop sequence">;
+def warn_omp_redundant_fusion : Warning<
+  "looprange clause selects a single loop, resulting in redundant fusion">,
+  InGroup<OpenMPClauses>;
```
I don't see all these messages uses

rofirrim replied:
`clang/lib/Sema/SemaOpenMP.cpp`

alexey-bataev commented on clang/lib/CodeGen/CGExpr.cpp:
```diff
@@ -3256,9 +3256,8 @@ LValue CodeGenFunction::EmitDeclRefLValue(const DeclRefExpr *E) {
 
     // No other cases for now.
-    } else {
+    } else
```
Restore braces here

Meinersbur commented:
> Maybe there is no need to recursively represent all the nested transformation?

I don't see a need to represent the entire loop structure. The only info that I think is potentially needed is the nest depth for transformations that generate a loop nest (e.g. tile), and for transformations that generate loop sequences it is the number of loops in that sequence (e.g. `split` for index-set-splitting, and by construction also every construct with a `looprange` clause).

The only use case for 2130117f92e51df73ac8c4b7e37f7f89178a89f2 was to identify whether -- while being in a non-fully instantiated template context -- there are additional generated loops or not. Within a dependend context `getTransformedStmt()` may return null because e.g. arguments of the `looprange()` clause might be template parameters. 
```c++
      Stmt *TransformedStmt = Dir->getTransformedStmt();
      if (!TransformedStmt) {
        unsigned NumGeneratedLoops = Dir->getNumGeneratedLoops();
        if (NumGeneratedLoops == 0) {
          // May happen if the loop transformation does not result in a
          // generated loop (such as full unrolling).
          break;
        }
        if (NumGeneratedLoops > 0) {
          // The loop transformation construct has generated loops, but these
          // may not have been generated yet due to being in a dependent
          // context.
          return true;
        }
      }
```

Outside of a template context `getTransformedStmt()` returns the fully expanded loops which contain the nesting information that can be analyzed by recursing into it, so no separate generated loop info should be necessary. Eventually the template will be used in a instantiated context (or it's completly unused) where it can be analyzed. Until then, all analysis can be defererred (which ` return true;` does). 

Since it is typically easily obtainable (e.g. by the number of items in a `sizes` clause for tile), the number of generated loops can be used for additional semantic checking even in a templated (/dependent) context. This is what the `break` above does: Since we know there are no further loops even in an instantiated context, we do not need to defer semantic checks until later, and analyze the loop nesting we already have.

I imagine those numbers could be used for better diagnostic output in general, e.g. when using `collapse(2)` on a `#pragma omp unroll partial`, hinting the user that the transformation does not produce a sufficient number of loops, instead of pointing out that the unrolled loop body is not a valid canonical loop[^1]. However, I think tracking the entire loop nest for better diagnostic messages is overblown, and may not even be accurate. For instance:
```c
for (int i = 1; i < 20; i+=1)
  #pragma omp tile sizes(i)
  for (int j = 1; j < n; j+=1)
```
Just recording the loop nesting of this may result in `gls = [ 3 ]`. Ovever, the tile size depends on the outer loop which the OpenMP restrictions on canonical loop nests do not allow. One may track the dependencies through all the clauses and loop bounds, but I think it is sufficient to let `checkOpenMPLoop` diagnose all those restrictions once `getTransformedStmt()` is guarenteed to be available.

[^1]: In the OpenMP spec the unrolled body is never a valid canonical loop, even with `partial(1)`, nor a loop sequence, a concept that was added after unroll. I am not sure what the use case would be for considering the unrolled loop body to be a canonical loop sequence.

rofirrim commented:
Hi @Meinersbur, thanks for the detailed answer.

> Just recording the loop nesting of this may result in gls = [ 3 ]. Ovever, the tile size depends on the outer loop which the OpenMP restrictions on canonical loop nests do not allow.

Right, this breaks my suggested bottom-up analysis. Thanks for this counterexample.

> One may track the dependencies through all the clauses and loop bounds, but I think it is sufficient to let `checkOpenMPLoop` diagnose all those restrictions once `getTransformedStmt()` is guarenteed to be available.

I understand this is predicated to the availability of the transformation in `Sema`, but this may not happen if we defer the transformation to `OpenMPIRBuilder` in `CodeGen`. For these cases I presume we cannot emit reliable diagnostics for the validity of the canonical loop nest. Is my understanding correct?

Anyways, no need to complicate matters further for this PR :-)

You also say:

> In the OpenMP spec the unrolled body is never a valid canonical loop, even with partial(1), nor a loop sequence, a concept that was added after unroll. I am not sure what the use case would be for considering the unrolled loop body to be a canonical loop sequence

Then I'm confused by this wording here in OpenMP 6.0, Section 11.9 Page 382, in particular the last statement seems to suggest that "presence of the partial clause == the resulting transform is a canonical loop nest". Is this a bug in the spec or I'm interpreting it incorrectly?

<img width="841" height="113" alt="image" src="https://github.com/user-attachments/assets/abf250c4-41fc-4c95-9b8b-86ba1833a31e" />

Thanks!

Meinersbur commented:
> I understand this is predicated to the availability of the transformation in `Sema`, but this may not happen if we defer the transformation to `OpenMPIRBuilder` in `CodeGen`. For these cases I presume we cannot emit reliable diagnostics for the validity of the canonical loop nest. Is my understanding correct?

Yes, in that case, either error diagnostic would be deferred to CodeGen (not ideal, but some diagnostics CodeGen are only ermitted in CodeGen), or Semantics would do additional analysis. The `-fopenmp-enable-irbuilder` indented to do that. It introduced a new AST node `OMPCanonicalLoop` that would make such semantic analysis significantly easier. 



> Then I'm confused by this wording here in OpenMP 6.0, Section 11.9 Page 382, in particular the last statement seems to suggest that "presence of the partial clause == the resulting transform is a canonical loop nest". Is this a bug in the spec or I'm interpreting it incorrectly?

This should mean that the partially unrolled loop is a canonical loop nest (of depth one), but this does not refer to what is in the loop body.

I think we have wording in the latest spec that combining a `partial(1)` unrolled loop together with a loop in the body cannot be used as a doacross-loop, but I am not sure about using it as a depth-2 canonical loop in other contexts. In any case, `partial(1)` is a such a unusual corner case that I would not bother about it.

rofirrim commented:
Thanks for the clarification @Meinersbur !

rofirrim commented on clang/include/clang/Sema/SemaOpenMP.h:
```diff
@@ -1479,7 +1492,109 @@ class SemaOpenMP : public SemaBase {
+      SmallVectorImpl<Stmt *> &ForStmts,
+      SmallVectorImpl<SmallVector<Stmt *>> &OriginalInits,
+      SmallVectorImpl<SmallVector<Stmt *>> &TransformsPreInits,
+      SmallVectorImpl<SmallVector<Stmt *>> &LoopSequencePreInits,
```
This is now grouped in two structs, one struct for the whole loop sequence and one struct for each loop in it.

Meinersbur commented on clang/test/OpenMP/fuse_ast_print.cpp:
```diff
@@ -0,0 +1,400 @@
+    }
+
+}
+
+
+
+
```
[nit] empty lines

alexey-bataev commented on clang/lib/AST/StmtOpenMP.cpp:
```diff
@@ -508,6 +512,43 @@ OMPInterchangeDirective::CreateEmpty(const ASTContext &C, unsigned NumClauses,
+  // not match the total number of generated loops
+  // Example:
+  // Before fusion:
+  //   for (int i = 0; i < N; ++i)
+  //     for (int j = 0; j < M; ++j)
+  //       A[i][j] = i + j;
+  //
+  //   for (int k = 0; k < P; ++k)
+  //     B[k] = k * 2;
+  // Here, NumLoopNests = 2, but NumLoops = 3.
```
Yep, not necessary to do it here, better in separate patch(es)

alexey-bataev commented on clang/include/clang/Sema/SemaOpenMP.h:
```diff
@@ -1479,7 +1492,85 @@ class SemaOpenMP : public SemaBase {
+    /// Number of canonical loops, including nested.
+    unsigned NumLoops = 0;
+    /// For each loop results of the analysis.
+    std::vector<LoopAnalysis> Loops;
```
SmallVector

rofirrim replied:
I had to use `SmallVector<LoopAnalysis, 2>` or a static assertion triggers about `LoopAnalysis` being too large.

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -14206,10 +14211,48 @@ StmtResult SemaOpenMP::ActOnOpenMPTargetTeamsDistributeSimdDirective(
 }
 
+/// Overloaded base case function
+template <typename T, typename F> static bool tryHandleAs(T *, F &&) {
+  return false;
+}
+
+/// Tries to recursively cast `Type` to one of the given types and invokes
+/// `Func` if successful.
+///
+/// \tparam Class The first type to check.
+/// \tparam Rest The remaining types to check.
+/// \tparam T The base type of `Type`.
+/// \tparam F The callable type for the function to invoke upon a successful
+///           cast.
+/// \param Type The object to be checked.
+/// \param Func The function to invoke if `Type` matches `Class`.
+/// \return `true` if `Type` matched any type and `Func` was called, otherwise
+/// `false`.
+template <typename Class, typename... Rest, typename T, typename F>
+static bool tryHandleAs(T *Type, F &&Func) {
+  if (Class *C = dyn_cast<Class>(Type)) {
+    Func(C);
+    return true;
+  }
+  return tryHandleAs<Rest...>(Type, std::forward<F>(Func));
+}
+
+/// Updates OriginalInits by checking Transform against loop transformation
+/// directives and appending their pre-inits if a match is found.
+static void updatePreInits(OMPLoopBasedDirective *Transform,
+                           SmallVectorImpl<Stmt *> &PreInits) {
+  if (!tryHandleAs<OMPTileDirective, OMPUnrollDirective, OMPReverseDirective,
+                   OMPInterchangeDirective, OMPFuseDirective>(
+          Transform, [&PreInits](auto *Dir) {
+            appendFlattenedStmtList(PreInits, Dir->getPreInits());
+          }))
+    llvm_unreachable("Unhandled loop transformation");
+}
+
```
Better to add and start using in a separate patch

rofirrim commented:
This change now builds on top of:

- https://github.com/llvm/llvm-project/pull/155849
- https://github.com/llvm/llvm-project/pull/155848

I added a new class `OMPCanonicalLoopSequenceTransformationDirective` which inherits from a common class `OMPLoopTransformationDirective` which is also a base class of `OMPCanonicalLoopNestTransformationDirective` (what used to be called `OMPLoopTransformationDirective`). This class only contains the number of (top-level) loop nests generated along with a some convenience methods for consistency.

`OMPFuseDirective` now inherits from `OMPCanonicalLoopSequenceTransformationDirective` as I think it doesn't make sense to make it inherit from `OMPLoopBasedDirective` like it used in an earlier version of this patch.

I have also squashed the changes for the Fuse implementation because it was getting hard to handle.

Please @Meinersbur, @alexey-bataev let me know if the refactoring seems reasonable.

alexey-bataev commented on clang/include/clang/AST/StmtOpenMP.h:
```diff
@@ -955,31 +960,69 @@ class OMPLoopBasedDirective : public OMPExecutableDirective {
+
+  /// We need this because we cannot easily make OMPLoopTransformationDirective
+  /// a proper Stmt.
+  Stmt *S;
```
```suggestion
  Stmt *S = nullptr;
```

alexey-bataev commented on clang/include/clang/AST/StmtOpenMP.h:
```diff
@@ -955,31 +960,69 @@ class OMPLoopBasedDirective : public OMPExecutableDirective {
+  Stmt *getPreInits() const;
+
+  static bool classof(const Stmt *T) {
+    return isa<OMPCanonicalLoopNestTransformationDirective>(T) ||
+           isa<OMPCanonicalLoopSequenceTransformationDirective>(T);
```
```suggestion
    return isa<OMPCanonicalLoopNestTransformationDirective, OMPCanonicalLoopSequenceTransformationDirective>(T);
```

alexey-bataev commented on clang/include/clang/AST/StmtOpenMP.h:
```diff
@@ -6582,4 +6740,40 @@ class OMPAssumeDirective final : public OMPExecutableDirective {
+
+  static clang::OMPLoopTransformationDirective *doCast(clang::Stmt *T) {
+    if (auto *D =
+            dyn_cast<clang::OMPCanonicalLoopNestTransformationDirective>(T)) {
+      return static_cast<clang::OMPLoopTransformationDirective *>(D);
+    }
```
Drop braces

alexey-bataev commented on clang/include/clang/AST/StmtOpenMP.h:
```diff
@@ -6582,4 +6740,40 @@ class OMPAssumeDirective final : public OMPExecutableDirective {
+    }
+    if (auto *D =
+            dyn_cast<clang::OMPCanonicalLoopSequenceTransformationDirective>(
+                T)) {
+      return static_cast<clang::OMPLoopTransformationDirective *>(D);
+    }
```
Drop braces

eZWALT committed:
[Clang][OpenMP] Implement #pragma omp fuse

Meinersbur reviewed (commented):
There is still @alexey-bataev's [request](https://github.com/llvm/llvm-project/pull/139293/files#r2159465143) to separate OMPLoopCategory into a separate PR.

Otherwise, I have no change request to make.

rofirrim committed:
Remove OMPLoopCategory enum and Category field

With the rearranged AST hierarchy we don't need it. Storing the original
Stmt should be enough now.

This change also introduces a few helper functions so we can keep the
generation of the analysis and the use of the analysis in sync (this
was previously done through the Category field)

rofirrim commented:
> There is still @alexey-bataev's [request](https://github.com/llvm/llvm-project/pull/139293/files#r2159465143) to separate OMPLoopCategory into a separate PR.

Yep, not forgetting about this one.

@alexey-bataev I've managed to remove the enum because with the recent AST restructuring we don't need it anymore. I've rearranged some functions to keep things in sync between the analysis creation and the analysis usage.

rofirrim committed:
Update cindex.py

DeinAlptraum reviewed (commented):
Re: Python bindings (`cindex.py`): LGTM

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -15716,6 +15987,489 @@ StmtResult SemaOpenMP::ActOnOpenMPInterchangeDirective(
+  assert(SeqAnalysis.LoopSeqSize == SeqAnalysis.Loops.size());
+
+  // Handle clauses, which can be any of the following: [looprange, apply]
+  const OMPLoopRangeClause *LRC =
```
```suggestion
  const auto *LRC =
```

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -15716,6 +15987,489 @@ StmtResult SemaOpenMP::ActOnOpenMPInterchangeDirective(
+  // Select the type with the largest bit width among all induction variables
+  QualType IVType =
+      SeqAnalysis.Loops[FirstVal - 1].HelperExprs.IterationVarRef->getType();
+  for (unsigned I = FirstVal; I < LastVal; ++I) {
```
```suggestion
  for (unsigned I : seq<unsigned>(FirstVal, LastVal)) {
```

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -15716,6 +15987,489 @@ StmtResult SemaOpenMP::ActOnOpenMPInterchangeDirective(
+  //  original loop. The preinit structure must ensure that hidden variables
+  //  like '.omp.fuse.max' are still properly handled.
+  // Transformations that apply this concept: Loopranged Fuse, Split
+  if (!SeqAnalysis.LoopSequencePreInits.empty()) {
+    llvm::append_range(PreInits, SeqAnalysis.LoopSequencePreInits);
+  }
```
Drop braces. Why need a check here?

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -15716,6 +15987,489 @@ StmtResult SemaOpenMP::ActOnOpenMPInterchangeDirective(
+      // inits such as upper bounds...
+      SmallVector<Stmt *> &TransformPreInit =
+          SeqAnalysis.Loops[TransformIndex++].TransformsPreInits;
+      if (!TransformPreInit.empty())
+        llvm::append_range(PreInits, TransformPreInit);
```
Drop the check

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -15716,6 +15987,489 @@ StmtResult SemaOpenMP::ActOnOpenMPInterchangeDirective(
+  //   omp.fuse.max = max(omp.temp1, omp.temp0)
+
+  ExprResult MaxExpr;
+  // I is the true
```
What does it mean?

alexey-bataev commented on clang/lib/Serialization/ASTWriterStmt.cpp:
```diff
@@ -2487,6 +2487,19 @@ void ASTStmtWriter::VisitOMPInterchangeDirective(OMPInterchangeDirective *D) {
+void ASTStmtWriter::VisitOMPCanonicalLoopSequenceTransformationDirective(
+    OMPCanonicalLoopSequenceTransformationDirective *D) {
+  VisitStmt(D);
+  Record.writeUInt32(D->getNumLoopsInSequence());
```
Is Uint32 size enough here?

rofirrim replied:
The field is (and the getter yields) an `unsigned` type. So I'd say so.

alexey-bataev replied:
Better to avoid such a transformation for ints. Or use uint32_t explicitly in the code instead of the unsigned

rofirrim commented on clang/include/clang/Sema/SemaOpenMP.h:
```diff
@@ -1485,7 +1498,81 @@ class SemaOpenMP : public SemaBase {
+
+    // Convenience functions used when building LoopSequenceAnalysis.
+    static bool isRegularLoop(Stmt *S) {
+      return isa<clang::ForStmt, CXXForRangeStmt>(S);
```
We have a field called `ForStmt` which hides the class name. I will rename the field instead.

rofirrim committed:
Address reviewers comments

rofirrim committed:
Missed feedback item to address

rofirrim committed:
Remove unused attribute from OMPCanonicalLoopSequenceTransformationDirective

When I refactored the hierarchy I had the impression it would be used
but it is not.

Also remove a stray NumGeneratedLoopNests field in OMPFuseDirective that
I forgot to remove during the refactoring.

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -14433,29 +14465,268 @@ bool SemaOpenMP::checkTransformableLoopNest(
+    return false;
+
+  // Diagnose an empty loop sequence.
+  if (SeqAnalysis.LoopSeqSize <= 0) {
```
LoopSeqSize unsigned, just ==0 should be enough

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -15716,6 +15987,485 @@ StmtResult SemaOpenMP::ActOnOpenMPInterchangeDirective(
+        if (NeedsNewVD) {
+          VD = buildVarDecl(SemaRef, SourceLocation(), IVType, Name);
+          SemaRef.AddInitializerToDecl(VD, TransformedExpr, false);
+
```
Delete empty line here

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```suggestion
    for (unsigned I : seq<unsigned>(SeqAnalysis.LoopSeqSize)) {
```

rofirrim committed:
Address review comments

alexey-bataev reviewed (commented):
LG with a nit

rofirrim committed:
Use enumerators rather than hardcoded values
